challenge_demo/test/grammar_test.py

```diff
@@ -0,0 +1,130 @@
+        actions_target = actions_list_rule.options[0].conjuncts[0].name
+
+        cls.actions_rule = cls.parser.rules[actions_target]
+
+    def test_grammar(self):
+
+        test_results = {}
+        for option in self.actions_rule.options:
+
+            result_str = ""
+
+            # For each conjunct, we want to have a viable option
+            for conjunct in option.conjuncts:
+                result_str += self._resolve_conjunct(conjunct.name)
+            result_str = result_str.rstrip(" ")
+
+            # Now, we can parse the result string to get the action description
+            actions_definition = self.parser.parse_raw(
+                target=self.knowledge.grammar_target,
+                words=result_str,
+            )
+            self.assertIn(
+                "actions",
+                actions_definition,
+                "'actions' not in actions definition, don't know what to do"
+            )
+
+            # Test the action description with the task manager
+            try:
+                test_result = self.task_manager.set_up_state_machine(
+                    recipe=actions_definition["actions"],
+                )  # type: ConfigurationResult
+            except Exception as e:
+                test_result = ConfigurationResult()
+                test_result.message = "Configuration crashed: {}".format(e.message)
+            test_results[result_str] = TMTestResult(actions_definition["actions"], test_result)
+
+        failed_test_results = {}
+        for action, test_result in test_results.iteritems():
+            if test_result.config_result.succeeded:
+                print("Configuration of '{}' succeeded".format(action))
+            elif test_result.config_result.message and test_result.config_result.missing_field:
+                print("Configuration of '{}' did not succeed due to missing information ({}): {}".format(
+                    action, test_result.config_result.missing_field, test_result.config_result.message
+                ))
+            else:
+                failed_test_results[action] = test_result
+
+        error_str = "\n"
+        for action, test_result in failed_test_results.iteritems():
+            error_str += "\nConfiguration of action '{}' failed.\n\tRecipe: {}\n\tError: {}".format(
+                action,
+                test_result.recipe,
+                test_result.config_result.message,
+            )
+        self.assertFalse(failed_test_results, error_str)
+
+    @classmethod
+    def _resolve_conjunct(cls, target):
+        # type: (str) -> str
+        """
+        Recursive method that tries to resolve a conjunct, i.e., to find an option
+
+        :param target: target to look for
+        :return: option string
+        """
+        if target not in cls.parser.rules:
+            return target + " "
+
+        result = ""
+        rule = cls.parser.rules[target]
+        # For now, take the first available option
```

Comment:
What should this become?

Reply:
See 74976f4b5effd72b089bd6e1d333fc2a0dc98fc9